charms/knative-operator/src/charm.py

```diff
@@ -0,0 +1,174 @@
+            "summary": "knative-operator layer",
+            "description": "pebble config layer for knative-operator",
+            "services": {
+                self._operator_service: {
+                    "override": "replace",
+                    "summary": "entrypoint of the knative-operator image",
+                    "command": self._operator_service,
+                    "startup": "enabled",
+                    "environment": {
+                        "POD_NAME": self._name,
+                        "SYSTEM_NAMESPACE": self._namespace,
+                        "METRICS_DOMAIN": "knative.dev/operator",
+                        "CONFIG_LOGGING_NAME": "config-loggig",
+                        "CONFIG_OBSERVABILITY_NAME": "config-observability",
+                    },
+                }
+            },
+        }
+        return Layer(layer_config)
+
+    def _update_layer(self) -> None:
+        """Updates the Pebble configuration layer if changed."""
+        if not self._container.can_connect():
+            self.unit.status = MaintenanceStatus("Waiting for pod startup to complete")
+            return
+
+        # Get current config
+        current_layer = self._container.get_plan()
+        # Create a new config layer
+        new_layer = self._knative_operator_layer
+        if current_layer.services != new_layer.services:
+            self._container.add_layer(self._operator_service, new_layer, combine=True)
+            logging.info("Pebble plan updated with new configuration")
+        self._container.restart(self._operator_service)
+
+    def _apply_all_resources(self) -> None:
+        """Helper method to create Kubernetes resources."""
+        for filename in self._resources_files:
+            manifest = self.env.get_template(filename).render(self._context)
+            for obj in codecs.load_all_yaml(manifest):
+                self.lightkube_client.apply(obj)
+
+    def _on_install(self, event):
+        """Event handler for InstallEvent."""
+
+        # Update Pebble configuration layer if it has changed
+        self._update_layer()
+
+        # Patch/create Kubernetes resources
+        try:
+            self.unit.status = MaintenanceStatus("Applying resources")
+            self._apply_all_resources()
+        except ApiError as e:
+            logging.error(traceback.format_exc())
+            self.unit.status = BlockedStatus(
+                f"Applying resources failed with code {str(e.status.code)}."
+            )
+            if e.status.code == 403:
+                logging.error(
+                    "Received Forbidden (403) error when creating auth resources."
+                    "This may be due to the charm lacking permissions to create"
+                    "cluster-scoped resources."
+                    "Charm must be deployed with --trust"
+                )
+                event.defer()
+                return
+        else:
+            self.unit.status = ActiveStatus()
```

Comment:
@DomFleischmann not sure what you mean by "I think in general the right way of checking if pebble has been started succesfully is by using [health checks](https://juju.is/docs/sdk/interact-with-pebble#heading--health-checks).", since we have the `PebbleReadyEvent`. I think @ca-scribner's comment is more related to checking that the restart actually succeeded before setting an `Active` status, and for that I think we can rely on pebble raising an `APIError` (I added some exception handling in my latest commit for that) or use pebble.replan() and handle `ChangeError` exceptions. I'm liking the latter, but I've to investigate more, as the examples I've seen so far do either.

Reply:
Sorry you are right I misworded that, it's not "if pebble has started successfully" but "if the **pebble layer/service** has started succesfully". I do think that handling the exceptions you mention is a start for this PR. But I would like us to revisit the health check feature in the future.